File: crates/riptide-api/src/handlers/crawl.rs
use crate::errors::ApiError;
use crate::facades::CrawlHandlerFacade;
use crate::models::{CrawlBody, CrawlResponse};
use crate::context::ApplicationContext;
use crate::telemetry_config::extract_trace_context;
use crate::validation::validate_crawl_request;
use axum::{extract::State, http::HeaderMap, Json};
use opentelemetry::trace::SpanKind;
use riptide_events::{BaseEvent, EventSeverity};
use std::time::Instant;
use tracing::{debug, info, warn, Span};

/// Batch crawl endpoint for processing multiple URLs concurrently.
///
/// This endpoint processes multiple URLs through the complete fetch->gate->extract pipeline:
/// 1. Validates input URLs and options
/// 2. Creates a pipeline orchestrator with appropriate settings
/// 3. Executes crawling concurrently while respecting rate limits
/// 4. Returns comprehensive results with statistics
///
/// Supports various crawl options including caching strategies, concurrency limits,
/// and extraction modes.
#[tracing::instrument(
    name = "crawl_handler",
    skip(state, body, headers),
    fields(
        http.method = "POST",
        http.route = "/crawl",
        url_count = body.urls.len(),
        cache_mode = ?body.options.as_ref().map(|o| &o.cache_mode),
        use_spider = ?body.options.as_ref().and_then(|o| o.use_spider),
        otel.kind = ?SpanKind::Server,
        otel.status_code
    )
)]
pub async fn crawl(
    State(state): State<ApplicationContext>,
    headers: HeaderMap,
    Json(body): Json<CrawlBody>,
) -> Result<Json<CrawlResponse>, ApiError> {
    let start_time = Instant::now();

    // Extract trace context from headers for distributed tracing (TELEM-004)
    if let Some(_parent_context) = extract_trace_context(&headers) {
        debug!("Trace context extracted from request headers");
    }

    // Record custom span attributes (TELEM-003)
    let current_span = Span::current();
    current_span.record("url_count", body.urls.len());
    if let Some(ref opts) = body.options {
        current_span.record("cache_mode", opts.cache_mode.as_str());
        current_span.record("concurrency", opts.concurrency);
        current_span.record("use_spider", opts.use_spider.unwrap_or(false));
    }

    info!(
        url_count = body.urls.len(),
        cache_mode = body.options.as_ref().map(|o| &o.cache_mode),
        "Received crawl request"
    );

    // Emit crawl start event
    let mut start_event = BaseEvent::new("crawl.started", "api.crawl_handler", EventSeverity::Info);
    start_event.add_metadata("url_count", &body.urls.len().to_string());
    if let Some(ref opts) = body.options {
        start_event.add_metadata("cache_mode", &opts.cache_mode.to_string());
        start_event.add_metadata("concurrency", &opts.concurrency.to_string());
    }
    if let Err(e) = state.event_bus.emit(start_event).await {
        warn!(error = %e, "Failed to emit crawl start event");
    }

    // Validate the request
    validate_crawl_request(&body)?;

    // Use provided options or defaults
    let options = body.options.unwrap_or_default();

    // Create facade for business logic
    let facade = CrawlHandlerFacade::new(state.clone());

    // Check if spider mode is requested and route accordingly
    let response = if options.use_spider.unwrap_or(false) {
        info!("Spider mode requested, routing to spider crawl");
        facade.crawl_spider_mode(&body.urls, &options).await?
    } else {
        // Execute batch crawl through facade
        facade.crawl_batch(&body.urls, options).await?
    };

    // Record success metrics in span (TELEM-003)
    let elapsed_ms = start_time.elapsed().as_millis() as u64;
    current_span.record("otel.status_code", "OK");
    current_span.record("http.status_code", 200);
    current_span.record("successful_count", response.successful);
    current_span.record("failed_count", response.failed);
    current_span.record("cache_hits", response.from_cache);
    current_span.record("cache_hit_rate", response.statistics.cache_hit_rate);
    current_span.record("total_time_ms", elapsed_ms);

    info!(
        total_urls = body.urls.len(),
        successful = response.successful,
        failed = response.failed,
        cache_hits = response.from_cache,
        total_time_ms = elapsed_ms,
        "Crawl request completed"
    );

    // Emit crawl completion event
    let mut complete_event =
        BaseEvent::new("crawl.completed", "api.crawl_handler", EventSeverity::Info);
    complete_event.add_metadata("total_urls", &body.urls.len().to_string());
    complete_event.add_metadata("successful", &response.successful.to_string());
    complete_event.add_metadata("failed", &response.failed.to_string());
    complete_event.add_metadata("cache_hits", &response.from_cache.to_string());
    complete_event.add_metadata("duration_ms", &start_time.elapsed().as_millis().to_string());
    complete_event.add_metadata(
        "cache_hit_rate",
        &format!("{:.2}", response.statistics.cache_hit_rate),
    );
    if let Err(e) = state.event_bus.emit(complete_event).await {
        warn!(error = %e, "Failed to emit crawl completion event");
    }

    // Record metrics for crawl request
    state.record_http_request("POST", "/crawl", 200, start_time.elapsed().as_secs_f64());

    Ok(Json(response))
}

File: sdk/python/riptide_sdk/endpoints/crawl.py
"""
Crawl API endpoint implementation

Provides batch crawling operations with support for caching, concurrency,
and various extraction modes.
"""

from typing import List, Optional
import httpx

from ..models import CrawlResponse, CrawlOptions
from ..exceptions import APIError, ValidationError


class CrawlAPI:
    """API for batch crawl operations"""

    def __init__(self, client: httpx.AsyncClient, base_url: str):
        """
        Initialize CrawlAPI

        Args:
            client: Async HTTP client
            base_url: Base URL for the API
        """
        self.client = client
        self.base_url = base_url

    async def batch(
        self,
        urls: List[str],
        options: Optional[CrawlOptions] = None,
    ) -> CrawlResponse:
        """
        Perform batch crawl of multiple URLs

        Args:
            urls: List of URLs to crawl
            options: Optional crawl options

        Returns:
            CrawlResponse with results and statistics

        Raises:
            ValidationError: If URLs are invalid
            APIError: If the API returns an error

        Example:
            >>> result = await client.crawl.batch(
            ...     ["https://example.com", "https://example.org"],
            ...     options=CrawlOptions(concurrency=10)
            ... )
            >>> print(f"Successful: {result.successful}/{result.total_urls}")
        """
        if not urls:
            raise ValidationError("URLs list cannot be empty")

        if len(urls) > 100:
            raise ValidationError("Maximum 100 URLs per batch request")

        # Validate URLs
        for url in urls:
            if not url.startswith(("http://", "https://")):
                raise ValidationError(f"Invalid URL: {url}")

        # Build request body
        body = {"urls": urls}
        if options:
            body["options"] = options.to_dict()

        # Make request
        response = await self.client.post(
            f"{self.base_url}/api/v1/crawl",
            json=body,
        )

        if response.status_code != 200:
            error_data = response.json() if response.text else {}
            raise APIError(
                message=error_data.get("error", {}).get("message", "Crawl failed"),
                status_code=response.status_code,
                response_data=error_data,
            )

        return CrawlResponse.from_dict(response.json())

    async def single(
        self,
        url: str,
        options: Optional[CrawlOptions] = None,
    ) -> CrawlResponse:
        """
        Crawl a single URL (convenience wrapper around batch)

        Args:
            url: URL to crawl
            options: Optional crawl options

        Returns:
            CrawlResponse with single result

        Example:
            >>> result = await client.crawl.single("https://example.com")
            >>> doc = result.results[0].document
        """
        return await self.batch([url], options)
